### ISSUE TITLE
Why the performance deteriorates after fine-tuning?

### ISSUE THREAD

**Issue by HDUyiming** [Asad-Ismail/Grounding-Dino-FineTuning#21]
**I ran the fine-tuning work of the fashion_dataset_subset dataset, but I found that the performance of the model was getting worse and worse.
Could the author please tell me why fine-tuning doesn't work?**
The performance of eopch-5 is 34.2 mAP50, epoch-70 is 27.5 mAP50, epoch-150 is 6.5 mAP50,....

Currently, I wrote a json script converted from cvs to COCO format:
```
import csv
import json
from collections import defaultdict

def csv_to_coco(csv_path, output_json_path):
    # 初始化COCO格式的数据结构
    coco_data = {
        "images": [],
        "annotations": [],
        "categories": []
    }
    
    # 读取CSV文件并组织数据
    with open(csv_path, 'r') as csv_file:
        csv_reader = csv.DictReader(csv_file)
        
        # 用于跟踪图像和类别ID
        image_id_map = {}
        category_id_map = {}
        next_image_id = 1
        next_category_id = 1
        next_annotation_id = 1
        
        for row in csv_reader:
            image_name = row['image_name']
            
            # 处理图像信息
            if image_name not in image_id_map:
                image_id_map[image_name] = next_image_id
                coco_data["images"].append({
                    "file_name": image_name,
                    "id": next_image_id,
                    "width": int(row['width']),
                    "height": int(row['height'])
                })
                next_image_id += 1
            
            # 处理类别信息
            label_name = row['label_name']
            if label_name not in category_id_map:
                category_id_map[label_name] = next_category_id
                coco_data["categories"].append({
                    "id": next_category_id,
                    "name": label_name,
                    "supercategory": "object"
                })
                next_category_id += 1
            
            # 处理标注信息
            bbox = [
                float(row['bbox_x']),
                float(row['bbox_y']),
                float(row['bbox_width']),
                float(row['bbox_height'])
            ]
            area = float(row['bbox_width']) * float(row['bbox_height'])
            
            coco_data["annotations"].append({
                "id": next_annotation_id,
                "image_id": image_id_map[image_name],
                "category_id": category_id_map[label_name],
                "segmentation": [],
                "area": area,
                "bbox": bbox,
                "iscrowd": 0
            })
            next_annotation_id += 1
    
    # 写入JSON文件
    with open(output_json_path, 'w') as json_file:
        json.dump(coco_data, json_file, indent=2)
    
    print(f"Successfully converted CSV to COCO JSON. Saved to {output_json_path}")

if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description='Convert CSV annotations to COCO JSON format')
    parser.add_argument('--csv_path', type=str, required=True, help='Path to input CSV file')
    parser.add_argument('--output_json', type=str, required=True, help='Path to output JSON file')
    
    args = parser.parse_args()
    
    csv_to_coco(args.csv_path, args.output_json)
```
Next, I wrote a program that could test the performance of the fine-tuned model:
```
import argparse
import os
import sys
import time

import numpy as np
import torch
import torch.nn as nn
from torch.utils.data import DataLoader

from groundingdino.models import build_model
import groundingdino.datasets.transforms as T
from groundingdino.util import box_ops, get_tokenlizer
from groundingdino.util.misc import clean_state_dict, collate_fn
from groundingdino.util.slconfig import SLConfig
from groundingdino.util.inference import load_model as load_inference_model
from config import ConfigurationManager, DataConfig, ModelConfig

import torchvision
from groundingdino.util.vl_utils import build_captions_and_token_span, create_positive_map_from_span
from groundingdino.datasets.cocogrounding_eval import CocoGroundingEvaluator


class CocoDetection(torchvision.datasets.CocoDetection):
    def __init__(self, img_folder, ann_file, transforms):
        super().__init__(img_folder, ann_file)
        self._transforms = transforms

    def __getitem__(self, idx):
        img, target = super().__getitem__(idx)  # target: list

        w, h = img.size
        boxes = [obj["bbox"] for obj in target]
        boxes = torch.as_tensor(boxes, dtype=torch.float32).reshape(-1, 4)
        boxes[:, 2:] += boxes[:, :2]  # xywh -> xyxy
        boxes[:, 0::2].clamp_(min=0, max=w)
        boxes[:, 1::2].clamp_(min=0, max=h)
        # 过滤无效的框
        keep = (boxes[:, 3] > boxes[:, 1]) & (boxes[:, 2] > boxes[:, 0])
        boxes = boxes[keep]

        target_new = {}
        image_id = self.ids[idx]
        target_new["image_id"] = image_id
        target_new["boxes"] = boxes
        target_new["orig_size"] = torch.as_tensor([int(h), int(w)])

        if self._transforms is not None:
            img, target = self._transforms(img, target_new)

        return img, target


class PostProcessCocoGrounding(nn.Module):
    """将模型输出转换为COCO API预期的格式"""

    def __init__(self, num_select=300, coco_api=None, tokenlizer=None) -> None:
        super().__init__()
        self.num_select = num_select

        assert coco_api is not None
        category_dict = coco_api.dataset['categories']
        cat_list = [item['name'] for item in category_dict]
        captions, cat2tokenspan = build_captions_and_token_span(cat_list, True)
        tokenspanlist = [cat2tokenspan[cat] for cat in cat_list]
        positive_map = create_positive_map_from_span(
            tokenlizer(captions), tokenspanlist)  # 80, 256. normed

        # 构建标准COCO类别ID映射
        self.id_map = {}
        for i, cat in enumerate(category_dict):
            self.id_map[i] = cat['id']

        # 构建从label_id到pos_map的映射
        new_pos_map = torch.zeros((max(self.id_map.values()) + 1, 256))
        for k, v in self.id_map.items():
            new_pos_map[v] = positive_map[k]
        self.positive_map = new_pos_map

    @torch.no_grad()
    def forward(self, outputs, target_sizes, not_to_xyxy=False):
        """执行计算
        参数:
            outputs: 模型的原始输出
            target_sizes: 维度为[batch_size x 2]的张量，包含批次中每个图像的大小
        """
        num_select = self.num_select
        out_logits, out_bbox = outputs['pred_logits'], outputs['pred_boxes']

        # pos map to logit
        prob_to_token = out_logits.sigmoid()  # bs, 100, 256
        pos_maps = self.positive_map.to(prob_to_token.device)
        # (bs, 100, 256) @ (91, 256).T -> (bs, 100, 91)
        prob_to_label = prob_to_token @ pos_maps.T

        assert len(out_logits) == len(target_sizes)
        assert target_sizes.shape[1] == 2

        prob = prob_to_label
        topk_values, topk_indexes = torch.topk(
            prob.view(out_logits.shape[0], -1), num_select, dim=1)
        scores = topk_values
        topk_boxes = topk_indexes // prob.shape[2]
        labels = topk_indexes % prob.shape[2]

        if not_to_xyxy:
            boxes = out_bbox
        else:
            boxes = box_ops.box_cxcywh_to_xyxy(out_bbox)

        boxes = torch.gather(
            boxes, 1, topk_boxes.unsqueeze(-1).repeat(1, 1, 4))

        # 从相对[0, 1]转换到绝对[0, height]坐标
        img_h, img_w = target_sizes.unbind(1)
        scale_fct = torch.stack([img_w, img_h, img_w, img_h], dim=1)
        boxes = boxes * scale_fct[:, None, :]

        results = [{'scores': s, 'labels': l, 'boxes': b}
                  for s, l, b in zip(scores, labels, boxes)]

        return results


def load_finetuned_model(model_config, use_lora=False, device="cuda"):
    """加载微调后的模型"""
    model = load_inference_model(model_config, use_lora)
    model = model.to(device)
    model = model.eval()
    return model


def main(args):
    # 加载配置
    data_config, model_config, training_config = ConfigurationManager.load_config(args.config_file)
    
    # 构建模型
    model = load_finetuned_model(model_config, training_config.use_lora, args.device)
    
    # 构建数据加载器
    transform = T.Compose(
        [
            T.RandomResize([800], max_size=1333),
            T.ToTensor(),
            T.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
        ]
    )
    dataset = CocoDetection(
        args.image_dir, args.anno_path, transforms=transform)
    data_loader = DataLoader(
        dataset, batch_size=1, shuffle=False, num_workers=args.num_workers, collate_fn=collate_fn)
    
    # 构建后处理器
    tokenlizer = get_tokenlizer.get_tokenlizer("bert-base-uncased")
    postprocessor = PostProcessCocoGrounding(
        coco_api=dataset.coco, tokenlizer=tokenlizer)
    
    # 构建评估器
    evaluator = CocoGroundingEvaluator(
        dataset.coco, iou_types=("bbox",), useCats=True)
    
    # 构建文本提示
    if args.text_prompt:
        caption = args.text_prompt
    else:
        # 如果未提供文本提示，则使用COCO类别
        category_dict = dataset.coco.dataset['categories']
        cat_list = [item['name'] for item in category_dict]
        caption = " . ".join(cat_list) + ' .'
    print("输入文本提示:", caption)
    
    # 运行推理
    start = time.time()
    for i, (images, targets) in enumerate(data_loader):
        # 获取图像和文本提示
        images = images.tensors.to(args.device)
        bs = images.shape[0]
        input_captions = [caption] * bs
        
        # 送入模型
        outputs = model(images, captions=input_captions)
        
        orig_target_sizes = torch.stack(
            [t["orig_size"] for t in targets], dim=0).to(images.device)
        results = postprocessor(outputs, orig_target_sizes)
        cocogrounding_res = {
            target["image_id"]: output for target, output in zip(targets, results)}
        evaluator.update(cocogrounding_res)
        
        if (i+1) % 30 == 0:
            used_time = time.time() - start
            eta = len(data_loader) / (i+1e-5) * used_time - used_time
            print(
                f"已处理 {i}/{len(data_loader)} 张图像。用时: {used_time:.2f}s，预计剩余时间: {eta:.2f}s")
    
    evaluator.synchronize_between_processes()
    evaluator.accumulate()
    evaluator.summarize()
    
    print("最终结果:", evaluator.coco_eval["bbox"].stats.tolist())
    
    # 保存详细指标到文件
    if args.output_file:
        metrics = {}
        metrics["AP"] = evaluator.coco_eval["bbox"].stats[0].item()  # AP @IoU=0.50:0.95
        metrics["AP50"] = evaluator.coco_eval["bbox"].stats[1].item()  # AP @IoU=0.50
        metrics["AP75"] = evaluator.coco_eval["bbox"].stats[2].item()  # AP @IoU=0.75
        metrics["APs"] = evaluator.coco_eval["bbox"].stats[3].item()  # AP for small objects
        metrics["APm"] = evaluator.coco_eval["bbox"].stats[4].item()  # AP for medium objects
        metrics["APl"] = evaluator.coco_eval["bbox"].stats[5].item()  # AP for large objects
        
        import json
        with open(args.output_file, 'w') as f:
            json.dump(metrics, f, indent=4)
        print(f"评估指标已保存到: {args.output_file}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        "评估Fine-tuned Grounding DINO在COCO上的性能", add_help=True)
    # 加载模型
    parser.add_argument("--config_file", "-c", type=str,
                      required=True, help="配置文件路径")
    parser.add_argument("--device", type=str, default="cuda",
                      help="运行设备 (默认: cuda)")
    
    # 后处理
    parser.add_argument("--num_select", type=int, default=300,
                      help="topk选择的数量")
    
    # COCO信息
    parser.add_argument("--anno_path", type=str,
                      required=True, help="COCO标注路径")
    parser.add_argument("--image_dir", type=str,
                      required=True, help="COCO图像目录")
    parser.add_argument("--num_workers", type=int, default=4,
                      help="数据加载器的工作线程数")
    
    # 文本提示
    parser.add_argument("--text_prompt", type=str, default=None,
                      help="用于测试的文本提示，用点号分隔类别，例如 'shirt .bag .pants'")
    
    # 输出
    parser.add_argument("--output_file", type=str, default=None,
                      help="保存评估指标的JSON文件路径")
    
    args = parser.parse_args()
    
    main(args)
```
Use the following command to perform the test

```
python eval_finetuned_groundingdino.py \
    --config_file configs/test_config.yaml \
    --anno_path path/to/coco/annotations/instances_val2017.json \
    --image_dir path/to/coco/val2017 \
    --text_prompt "shirt .bag .pants" \
    --output_file eval_results.json
```

**Comment by wzl1026** [comment]
请问你解决了吗，可以一起交流吗

**Comment by HDUyiming** [comment]
> 请问你解决了吗，可以一起交流吗

我解决了，作者REAMER写的和代码中配置文件不一致，按照README中的配置修改一下就可以了
`Run the train.py for training. We use a batch size of 8, a learning rate of 1e-5, and the AdamW optimizer. The model is trained for 100 epochs. See configs/train_config.yaml for detailed training configurations.`

**Comment by wzl1026** [comment]
可以加一个联系方式吗。这是我的18780123903.谢谢您

**Comment by wzl1026** [comment]
我还发现一个问题，微调后是否丧失了预训练功能
I also found a problem: whether the pre-training function is lost after fine-tuning.
